---
title: "Assignment 1: Dataset Selection and Initial Processing"
author: Evgeniya Gorobets
date: "`r format(Sys.time(), '%d %b %Y')`"
output:
  html_document:
    df_print: paged
---

## Preface

This R Notebook was developed by Evgeniya Gorobets as part of an assessment for 
2022 BCB420H: Computational Systems Biology, University of Toronto, Toronto, CA.
Specifically, this notebook was the final submitted product for Assignment 1 
of the course.

This notebook only includes the download, cleaning, and processing of my 
selected GEO dataset. To view the code used to explore GEO, evaluate various 
GSEs, and ultimately select this dataset, see the R Notebook located at 
https://github.com/bcb420-2022/Evgeniya_Gorobets/blob/main/Assignment1/ExploringGEO.Rmd.

## Setup

The code below was taken from the R Notebook I did as a pre-requisite to this 
notebook, titled "Choosing a GEO Dataset". You can view the original code 
[here](https://github.com/bcb420-2022/Evgeniya_Gorobets/blob/main/Assignment1/ExploringGEO.Rmd).

```{r}
# Install BiocManage and GEOquery if required
if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}
if (!requireNamespace("GEOquery", quietly = TRUE)) {
    BiocManager::install("GEOquery")
}
if (!requireNamespace("edgeR", quietly = TRUE)) {
    BiocManager::install("edgeR")
}


# Change working directory, if needed
if (getwd() == "/home/rstudio") {
  setwd("./projects/Assignment1")
}

# Set constants
GSE <- 'GSE129943'
DATA_DIR <- 'data'
```

<br/>

## Downloading the Data

Add some commentary here

```{r}
gseDir <- paste0(DATA_DIR, "/", GSE)

# Download the supplemental files for GSE129943 if they're not yet in data dir
if (!file.exists(gseDir)) {
  dataFiles <- GEOquery::getGEOSuppFiles(GSE, baseDir=DATA_DIR)
  filePath <- rownames(dataFiles)
} else {
  filePath <- paste0(gseDir, "/", list.files(path=gseDir))
}

# Load supplemental files as a data frame
gseData <- read.delim(filePath, header=TRUE, check.names=FALSE)
# The columns that contain count data (useful for downstream analysis)
countsCols <- c("N1H1_Pool_1", "N1H1_Pool_2", "N1H1_Pool_3", 
                "P1A8_Pool_1", "P1A8_Pool_2", "P1A8_Pool_3")
```

<br/>

## Cleaning the Data

According to the `edgeR` protocol, the first step in cleaning the data is to 
remove weakly expressed genes. Specifically, we should filter out any genes 
that have less than one read per million (CPM) in n or more of the samples 
(where n is the size of the smallest group or replicates). Each group of 
replicates in our dataset has 3 samples, so we set n=3.

```{r}
nGenes <- length(unique(gseData$Ensembl))
print(paste("Number of genes pre-filtering:", nGenes))  # 57,905

# Use edgeR to compute CPMs
cpms <- edgeR::cpm(gseData[, countsCols])
rownames(cpms) <- gseData$Ensembl

# Only keep genes with CPM > 1 in at least 3 of the samples
n <- 3
selectedGenes <- rowSums(cpms > 1) >= n
gseData <- gseData[selectedGenes, ]

nGenes <- length(unique(gseData$Ensembl))
print(paste("Number of genes post-filtering:", nGenes))  # 14,241
```
Since we have removed some genes, the our library size has changed. Therefore, 
we need to recompute the CPMs in our dataset.
```{r}
cpms <- edgeR::cpm(gseData[, countsCols])
```


The second step of cleaning the data is to check for any duplicate genes.

```{r}
duplicateGenes <- table(gseData$Ensembl) > 1
length(gseData$Ensembl[duplicateGenes])
```
Since there are no duplicate genes, then this section of our data cleaning is 
done.
<br/>

## Normalizing the Data

Before we begin normalizing our data, we will first visualize the distribution 
of the data using a histogram and a boxplot.

```{r}
# Get the log2 of the CPMS
logCPMs <- log2(cpms)

# Draw a boxplot representing the quartiles and outliers for each sample
boxplot(logCPMs, xlab="Samples", ylab="log2(CPM)", cex.axis = 0.75,
        main=paste("Log2 CPMs of Bulk RNASeq Samples from", GSE))

# Compute median log2(CPM) across all genes in all samples and plot it
mediansPerSample <- apply(logCPMs, 2, median)
gseMedian <- median(mediansPerSample)
abline(h=gseMedian, col="blue")

```
<!-- reread assignment requirements and consider removing this paragraph -->
When creating the boxplots, I noticed some warning about outliers (-Inf) in 
some of the boxplots not being drawn. This indicates that some genes still have 
CPM=0 in some samples. However, this is simply means that some genes have CPM=0
in some samples but CPM>=1 in at least 3 of the samples, indicating possible 
differential expression.

Next, we will plot the densities of the log2(CPM) for each sample.

```{r}
# slide 44
```


Which case is this of the four ? 
Which assumptions are we making ?
Based on these assumptions, we will use the TMM normalization method.

```{r}

```


Need to map distribution again

## Mapping Gene Identifiers


## Clean-Up

```{r}
# Clear workspace
rm(list = ls())
```

## References

Isserlin, R. (2022). Lecture 3 - Finding Expression Data. Quercus.
https://q.utoronto.ca/courses/248455/files/18120910?module_item_id=3210846

My original R notebook

